### PR TITLE
Use Func<Task>

### PR DESCRIPTION
**Pull request by benaadams**

### CODE REVIEW

File: Program.cs [hunk 1]
```diff
@@ -456,10 +456,10 @@ private static Task ContinuationChain(int depth)
         {
             var tasks = new Task[dop];
             var subCount = Math.Max(1L, count / dop);
+            Func<Task> body = () => testAsync(depth, subCount);
 
             for (var i = 0; i < dop; i++)
             {
-                Action body = () => testAsync(depth, subCount);
                 tasks[i] = dopFromPool ? Task.Run(body) : nonTpFactory.StartNew(body);
```
Comment (stephentoub) [inline Program.cs:463]: The StartNew call also needs to have .Unwrap() appended to it.

Comment (benaadams) [reply Program.cs:463]: Wondered why that version was still really really fast 😄

**Comment by stephentoub** [comment]
Now that you've fixed the tests, how is the new runtime comparing against the old in your runs?

**Comment by benaadams** [comment]
Its mostly improved and some not improved; but variable between runs. I think concurrency + randomness makes it tricky to measure 😄 

I did a memory allocation profile for both plaintext and our app and all the queue segment allocations are gone - these tests try to eliminate GC as a factor as much as they can; so reduced GC won't show.

Also merged the two task PRs and the Task tests are improved across the board.

**Comment by benaadams** [comment]
The QUWI outlier regressions (which change between runs) are much worse in the previous; and there are less of them in the new.

**Comment by stephentoub** [comment]
So all in all, better than before (and in some cases significantly), right?

**Comment by benaadams** [comment]
Yes and more consistent; doing more analysis though.

**Comment by benaadams** [comment]
A lot of time in Enqueue spinning

**Comment by stephentoub** [comment]
You could try removing the spinning and see how it affects things. There are two cases that result in spinning here: I expect one is very reasonable to remove the spinning from (when the race on the compare exchange is lost)... the other I'd be a bit more hesitant about, but should also be more rare. Regardless, we'd want to try on multiple sized machines before removing it. But give it a go.

**Comment by benaadams** [comment]
Hmm... something else weird, end up with 500+ threads (task manager agrees)
* 6 threadpool that execute code
* 1 Main, mostly waiting
* 1 Finalizer, not doing anything
* 500+ "Clr Worker" threads that just seem to be asleep and don't execute anything

Might have also been doing this previously, will check

**Comment by benaadams** [comment]
High thread count was happening before the changes

**Comment by stephentoub** [comment]
What's the call stack? You're sure they're not coming from the dedicated thread scheduler I added to your tests?

**Comment by benaadams** [comment]
Ahh... yes, that's the cause 😄 

Did you mean this SpinWait? https://github.com/benaadams/coreclr/commit/53a290f5566bd58039cc16ce80a3adb4ea79ef00

Also having good success with the additional padding in the commit; don't think having poor cache coherence between the items matters vs executing an item between them. (Different than the general use case of corefx's ConcurrentQueue)

**Comment by stephentoub** [comment]
> Did you mean this SpinWait?

Yes, but that is hit in two cases.  I was suggesting I think it could make sense to avoid the spinning entirely if you get the implicit else here https://github.com/benaadams/coreclr/commit/53a290f5566bd58039cc16ce80a3adb4ea79ef00#diff-a993dee28477af368dddfd4ac55437afR1079, but still have the spinning with SpinWait in case you hit the implicit else here:
https://github.com/benaadams/coreclr/commit/53a290f5566bd58039cc16ce80a3adb4ea79ef00#diff-a993dee28477af368dddfd4ac55437afR1089

> Also having good success with the additional padding in the commit; don't think having poor cache coherence between the items matters vs executing an item between them. (Different than the general use case of corefx's ConcurrentQueue)

Still makes me nervous.  For this to really matter, seems like you'd need to commonly have four or fewer items in the queue (on 64-bit), as there should already be contention between producers and producers, and between consumers and consumers, so this would really only be helping between producers and consumers, in which case the items they're taking from the beginning and end of the queue would need to be in the same cache line.  I don't know how often that actually occurs in non-microbenchmarks.  And adding this will cause all cases to need to load at least 4x more memory (and consume 4x more memory).  This is one where I'd really want to see a measurable and consistent impact on a real consumer, like the techempower benchmark, and not just in microbenchmarks, as I'm not convinced we have an accurate set of benchmarks to reflect real-world usage.

**Comment by benaadams** [comment]
> For this to really matter, seems like you'd need to commonly have four or fewer items in the queue

So head and tail clash? Wouldn't you also get clashes with simulations enqueues and dequeues? The double write will clash with the next 4 item's enqueue if they are happening cross core
```
_slots[slotsIndex].Item = x
Volatile.Write(ref _slots[slotsIndex].SequenceNumber
```
```
_slots[slotsIndex + 1].Item = x // same cache line
_slots[slotsIndex + 2].Item = x // same cache line
_slots[slotsIndex + 3].Item = x // same cache line
```

**Comment by stephentoub** [comment]
> The double write will clash with the next 4 item's enqueue if they are happening cross core

My point was that you would have already had a clash between, for example, producers in trying to read/update the head index, so I'm not sure you'll end up with significantly more contention in then writing the item.  It's of course possible, I just hypothesized it wouldn't be nearly as much.  If I'm wrong, happy to be proven as such.

**Comment by benaadams** [comment]
K, 16 Core (32 HT) the changes are definitely an improvement - but the global queue contention is huge vs less procs. Tasks using local queues switch behaviour with global.

4 cores
```
QUWI No Queues (TP)             10.139 M     9.482 M     6.468 M     8.071 M    10.455 M
- Depth    2                     6.965 M     7.025 M     6.108 M     7.065 M     7.073 M
- Depth   16                     7.879 M     7.221 M     8.144 M     8.202 M     8.576 M
- Depth   64                     8.255 M     7.913 M     8.516 M     8.652 M     8.784 M
- Depth  512                     8.939 M     8.693 M     8.928 M     8.687 M     8.921 M

QUWI No Queues                  10.425 M    10.394 M    10.663 M    10.183 M     6.626 M
- Depth    2                     5.976 M     5.062 M     3.449 M     6.366 M     4.394 M
- Depth   16                     7.177 M     6.364 M     8.988 M     8.071 M     4.416 M
- Depth   64                     9.213 M     9.239 M     9.082 M     8.824 M     4.603 M
- Depth  512                     9.352 M     9.332 M     9.272 M     9.103 M     4.312 M

SubTask Chain Return (TP)      662.611 k     1.590 M     4.638 M     4.565 M     4.663 M
- Depth    2                   707.060 k     1.841 M     4.354 M     4.567 M     4.762 M
- Depth   16                   991.552 k     2.385 M     5.080 M     5.205 M     5.352 M
- Depth   64                   674.003 k     2.517 M     5.172 M     5.136 M     5.365 M
- Depth  512                   942.504 k     2.488 M     4.981 M     5.269 M     5.463 M

SubTask Chain Return           670.382 k     1.656 M     4.379 M     4.261 M     2.486 M
- Depth    2                   711.646 k     1.910 M     4.438 M     4.526 M     3.071 M
- Depth   16                   916.746 k     2.362 M     4.951 M     5.035 M     3.034 M
- Depth   64                   649.490 k     2.460 M     4.959 M     5.181 M     3.254 M
- Depth  512                   904.034 k     2.411 M     4.842 M     4.978 M     3.259 M
```
16  (32 HT)
```
QUWI No Queues (TP)            413.808 k   410.068 k     1.694 M     1.792 M     1.789 M
- Depth    2                   729.126 k     1.649 M     1.534 M     1.636 M     1.666 M
- Depth   16                   588.991 k   795.239 k     1.296 M     1.384 M     1.538 M
- Depth   64                     1.129 M     1.234 M     1.553 M     1.517 M     1.573 M
- Depth  512                     1.583 M     1.568 M     1.524 M     1.526 M     1.593 M

QUWI No Queues                 394.238 k     1.062 M     1.565 M     1.142 M     1.073 M
- Depth    2                   735.666 k     1.655 M     1.491 M     1.166 M     1.102 M
- Depth   16                   780.140 k   660.359 k     1.280 M     1.600 M   614.769 k
- Depth   64                     1.180 M     1.231 M     1.560 M     1.560 M   878.993 k
- Depth  512                     1.604 M     1.533 M     1.562 M     1.557 M     1.434 M

SubTask Chain Return (TP)      394.676 k   423.496 k     3.637 M    10.421 M    13.722 M
- Depth    2                   393.058 k   427.897 k     4.403 M    10.525 M    14.634 M
- Depth   16                   369.727 k   419.207 k     3.822 M    12.007 M    14.301 M
- Depth   64                   356.568 k   415.433 k     4.904 M    13.153 M    13.640 M
- Depth  512                   350.689 k   415.466 k     4.198 M    13.380 M    15.013 M

SubTask Chain Return           394.268 k   424.518 k     3.062 M     7.122 M     5.964 M
- Depth    2                   391.377 k   428.171 k     2.516 M     7.237 M     3.311 M
- Depth   16                   368.727 k   417.539 k     4.668 M     8.497 M     2.065 M
- Depth   64                   360.350 k   414.291 k     3.272 M     7.798 M     2.320 M
- Depth  512                   349.105 k   407.087 k     3.448 M     8.855 M     2.469 M
```

**Comment by stephentoub** [comment]
> 16 Core (32 HT) the changes are definitely an improvement 

Not sure how to read these results:
- these were run with what set of changes?
- what do I compare them against?
- is it valid to compare the results across the machines? e.g. same kind/speed of processor, size/speed of caches, etc.?

**Comment by benaadams** [comment]
> > 16 Core (32 HT) the changes are definitely an improvement

> these were run with what set of changes?

Old vs new (current coreclr); so ignoring those as they are a clear win.

The results I'm showing are current coreclr on two different machines; 2.9 GHz 4 core vs 2.4 GHz 16(32) core 

For example changing to QUWI to `ThreadPoolGlobals.workQueue.Enqueue(tpcallBack, forceGlobal: false);`

```
                                  Serial          2x         16x         64x        512x
QUWI No Queues (TP)            382.748 k   383.552 k    15.992 M    16.817 M    16.972 M
- Depth    2                   402.615 k   633.205 k    17.383 M    17.203 M    17.181 M
- Depth   16                    11.077 M    14.795 M    17.630 M    17.280 M    17.632 M
- Depth   64                    12.036 M    16.831 M    17.020 M    17.037 M    17.802 M
- Depth  512                    11.747 M    14.991 M    16.254 M    15.765 M    16.490 M

QUWI No Queues                 382.986 k   400.529 k     1.138 M     1.101 M     1.063 M
- Depth    2                   402.698 k   577.917 k     2.169 M     2.178 M     1.257 M
- Depth   16                    12.349 M    14.903 M    14.462 M    11.391 M     1.044 M
- Depth   64                    12.674 M    14.961 M    14.265 M    14.417 M     1.560 M
- Depth  512                    12.793 M    13.610 M    14.237 M    14.300 M     5.326 M

SubTask Chain Return (TP)      392.281 k   418.430 k     4.346 M    10.874 M    13.394 M
- Depth    2                   390.732 k   423.373 k     4.005 M    11.624 M    13.795 M
- Depth   16                   368.232 k   412.879 k     5.987 M    12.419 M    14.313 M
- Depth   64                   359.371 k   411.998 k     5.129 M    12.772 M    14.654 M
- Depth  512                   348.127 k   405.058 k     5.201 M    11.147 M    15.121 M

SubTask Chain Return           390.009 k   416.183 k     2.556 M     7.865 M     1.307 M
- Depth    2                   390.034 k   421.502 k     3.464 M     8.186 M     4.046 M
- Depth   16                   368.276 k   410.153 k     2.576 M     6.183 M     2.653 M
- Depth   64                   360.104 k   409.409 k     4.545 M     8.583 M     2.059 M
- Depth  512                   349.370 k   405.387 k     4.887 M     8.927 M     2.654 M
```
Though it only makes a really big difference if queuing from a threadpool thread (as shown by then non-TP queuer)

Haven't tested with padding or spinning changes; seeing if can see hot spots first.

**Comment by benaadams** [comment]
Wonder if changing that from `true` to `false` for non-UnsafeQUWI would be a terrible thing to do? Probably niche areas where it would apply, but on 4 Cores:

Force global: false
```
                                  Serial          2x         16x         64x        512x
QUWI No Queues (TP)              4.883 M    11.497 M    20.644 M    21.607 M    21.529 M
- Depth    2                     8.404 M    12.549 M    20.430 M    21.256 M    21.119 M
- Depth   16                    15.982 M    17.698 M    21.033 M    21.035 M    20.424 M
- Depth   64                    19.005 M    19.891 M    23.452 M    23.664 M    22.362 M
- Depth  512                    20.067 M    20.794 M    23.221 M    22.786 M    22.965 M

QUWI No Queues                   9.797 M    10.171 M    10.457 M    10.197 M     5.997 M
- Depth    2                    16.026 M    16.246 M    15.948 M    13.223 M     6.084 M
- Depth   16                    19.289 M    19.402 M    18.444 M    15.533 M     5.919 M
- Depth   64                    19.666 M    19.260 M    18.904 M    16.192 M     6.020 M
- Depth  512                    20.191 M    20.251 M    18.962 M    16.067 M     5.859 M
```
Force global: true
```
                                  Serial          2x         16x         64x        512x
QUWI No Queues (TP)             10.137 M    10.634 M    10.653 M    10.273 M    10.327 M
- Depth    2                     8.058 M     8.047 M     8.066 M     8.078 M     8.011 M
- Depth   16                     8.603 M     8.559 M     8.600 M     8.556 M     8.563 M
- Depth   64                     8.611 M     8.729 M     8.688 M     8.657 M     8.692 M
- Depth  512                     8.813 M     8.662 M     8.744 M     8.751 M     8.628 M

QUWI No Queues                  10.270 M    10.666 M    10.540 M    10.563 M     6.587 M
- Depth    2                     8.184 M     8.043 M     7.726 M     7.912 M     4.755 M
- Depth   16                     8.756 M     8.528 M     8.974 M     8.244 M     4.280 M
- Depth   64                     9.172 M     9.237 M     9.073 M     8.441 M     4.595 M
- Depth  512                     9.202 M     9.331 M     9.052 M     9.362 M     5.367 M
```
Will look more into the global queue contention, specifically.

**Comment by benaadams** [comment]
Currently is using 50% of the CPU with kernel time; so I'd guess its in spin sleeps; but its also too chaotic to measure.
```
// Run 1
Testing 2,621,440 calls, with GCs after 262,144 calls.
Operations per second on 32 Cores
                                                                             Parallelism
                                  Serial          2x         16x         64x        512x
QUWI No Queues (TP)            415.462 k   499.604 k     2.170 M     2.507 M     2.441 M
- Depth    2                   614.530 k     2.092 M     1.798 M     1.787 M     2.117 M
- Depth   16                   668.259 k   818.005 k     1.301 M     1.755 M     1.823 M
- Depth   64                     1.082 M     1.062 M     1.664 M     1.805 M     1.806 M
- Depth  512                     1.817 M     1.823 M     1.818 M     1.819 M     1.837 M

// Run 2
QUWI No Queues (TP)            396.551 k   413.562 k     1.386 M     1.433 M     1.457 M
- Depth    2                   851.811 k     1.298 M     1.419 M     1.417 M     1.401 M
- Depth   16                     1.376 M     1.234 M     1.334 M     1.374 M     1.374 M
- Depth   64                     1.353 M     1.331 M     1.364 M     1.347 M     1.336 M
- Depth  512                     1.327 M     1.343 M     1.314 M     1.321 M     1.304 M

// Run 3
QUWI No Queues (TP)            395.859 k   441.895 k     1.867 M     1.887 M     1.882 M
- Depth    2                   890.309 k     1.649 M     1.747 M     1.750 M     1.736 M
- Depth   16                     1.592 M     1.478 M     1.617 M     1.581 M     1.582 M
- Depth   64                     1.580 M     1.546 M     1.529 M     1.504 M     1.501 M
- Depth  512                     1.578 M     1.515 M     1.483 M     1.472 M     1.446 M

// Run 4
QUWI No Queues (TP)            391.724 k   410.759 k     1.758 M     1.875 M     1.875 M
- Depth    2                   963.735 k     1.467 M     1.645 M     1.661 M     1.643 M
- Depth   16                     1.525 M     1.509 M     1.438 M     1.513 M     1.495 M
- Depth   64                     1.497 M     1.445 M     1.465 M     1.440 M     1.434 M
- Depth  512                     1.494 M     1.454 M     1.415 M     1.413 M     1.392 M
```
K, backed out specifically all the TP changes (including ConcurrentQueue, Spin.TryEnter) and kept everything else (so latest coreclr i.e removing finaly with empty try etc).

Uses 100% CPU
```
// Run 1
Testing 2,621,440 calls, with GCs after 262,144 calls.
Operations per second on 32 Cores
                                                                             Parallelism
                                  Serial          2x         16x         64x        512x
QUWI No Queues (TP)            387.874 k   413.791 k     1.462 M     2.048 M     2.833 M
- Depth    2                   427.150 k   500.186 k     2.347 M     2.605 M     2.607 M
- Depth   16                     2.410 M     2.408 M     2.438 M     2.494 M     2.519 M
- Depth   64                     2.408 M     2.408 M     2.525 M     2.392 M     2.470 M
- Depth  512                     2.317 M     2.307 M     2.473 M     2.489 M     2.577 M

// Run 2
QUWI No Queues (TP)            400.696 k   412.435 k     1.268 M     2.011 M     2.854 M
- Depth    2                   431.678 k   457.733 k     2.242 M     2.558 M     2.590 M
- Depth   16                     2.355 M     2.396 M     2.378 M     2.486 M     2.486 M
- Depth   64                     2.424 M     2.511 M     2.565 M     2.574 M     2.580 M
- Depth  512                     2.381 M     2.335 M     2.431 M     2.464 M     2.468 M

// Run 3
QUWI No Queues (TP)            400.769 k   410.686 k     1.595 M     2.025 M     2.636 M
- Depth    2                   431.062 k   446.927 k     2.505 M     2.620 M     2.735 M
- Depth   16                     2.508 M     2.525 M     2.515 M     2.631 M     2.655 M
- Depth   64                     2.533 M     2.602 M     2.553 M     2.672 M     2.815 M
- Depth  512                     2.641 M     2.570 M     2.648 M     2.611 M     2.636 M

// Run 4
QUWI No Queues (TP)            393.389 k   404.282 k     1.598 M     2.063 M     3.049 M
- Depth    2                   432.420 k   449.234 k     2.565 M     2.699 M     2.766 M
- Depth   16                     2.441 M     2.517 M     2.607 M     2.577 M     2.681 M
- Depth   64                     2.538 M     2.504 M     2.639 M     2.585 M     2.551 M
- Depth  512                     2.412 M     2.411 M     2.530 M     2.527 M     2.647 M
```
Again, chaotic, but more 2M numbers, and even a 3M.

I added the spin skip; and that takes it to 100% cpu and is an improvement, but am not convinced I'm doing it right. Added spin skip + the padding, and that was worse. Used a item struct wrapper for ConcurrentQueue to make it a concrete type, was a little better nothing huge.

Was doing a lot of before, after, before, after etc which is where I noticed the results weren't making sense. So went to recheck against the current coreclr as a baseline, only running the first set.

So not sure now, is difficult to compare when can't get consistent results. 

Known definites: Currently only uses 50% of 16/32 core machine. Previously used 100% (for most of the runs not all). Adding the spin skip takes it to 100% cpu for those ones also.

Will see what removing random (and all the TP changes) does across runs; will start with them all hitting the first queue rather than anything fancy - just looking to see if its possible to stabilise the results rather than any fairness.

**Comment by benaadams** [comment]
Looking at the flip side... is a gain of 50% of cpu for much less than half loss in throughput 😛

**Comment by benaadams** [comment]
No removing random, didn't help, might need to get the actions to do some work rather than just being empty.

**Comment by benaadams** [comment]
Though either way the spin skip will probably help; if there is less contention from doing more work, then it doesn't come into play; if there is more contention it sleeps less.

**Comment by benaadams** [comment]
Recheck assumptions part 2; not sure I trust the tests... Making sure they don't have cache conflicts themselves.

**Comment by benaadams** [comment]
Fixed the tests some. The multi-core deterioration bothers me more than the raw throughput

**Comment by benaadams** [comment]
Fixed the tests some. The multi-core deterioration at this point bothers me more than the raw throughput.

I understand why its happening; but as a user you'd expect to add more cores and get higher throughput rather than lower throughput.

**Comment by stephentoub** [comment]
> but as a user you'd expect to add more cores and get higher throughput rather than lower throughput.

The tests are doing close to 0 work, so more cores doesn't help, it just adds contention and makes the existing effectively serialized operation slower.  More cores would help if there was any actual work being performed per work item.

**Comment by benaadams** [comment]
True. Will add work...

**Comment by benaadams** [comment]
Yeah, I think its a victim of its own success, its faster which means it has more contention, which means slower

**Comment by benaadams** [comment]
Yeah, I think its a victim of its own success, its faster which means it has more contention, which means slower (for very low work items)

**Comment by benaadams** [comment]
Is significantly faster on a real world test; haven't managed to capture it well in a benchmark though 😞

**Comment by stephentoub** [comment]
> I think its a victim of its own success, its faster which means it has more contention, which means slower

This is also why the SpinOnce calls are there.  If there's no contention, SpinOnce won't be used.  If there is, it's called to help each thread back off a bit.  SpinOnce itself increases what it does as a backoff the more times it's called, with a mix of Thread.SpinWait, Thread.Yield, Thread.Sleep(0), and Thread.Sleep(1).

> haven't managed to capture it well in a benchmark though

Have you tried out the various ASP.NET benchmarks to see what kind of impact the new corelib has with the checked-in thread pool updates?

**Comment by benaadams** [comment]
> Have you tried out the various ASP.NET benchmarks to see what kind of impact the new corelib has with the checked-in thread pool updates?

I did a cursory local test on 4 cores; and was about at 7% gain, also not really valid test. Setting some servers up to do a proper multicore network test later today.

**Comment by benaadams** [comment]
> Have you tried out the various ASP.NET benchmarks to see what kind of impact the new corelib has with the checked-in thread pool updates?

I did a cursory local test on 4 cores; and was about at 7% gain, also not really valid test. Setting some servers up to do a proper multicore network test later today.

Also want to try SslStream etc with it; though need to work through how to combine corefx and coreclr in a happy way - a lot is in flux.

**Comment by stephentoub** [comment]
> Setting some servers up to do a proper multicore network test later today.

Cool, thanks!  Waiting with bated breath... :)

**Comment by benaadams** [comment]
I'm getting dreadful results 😞 However they aren't good before or after (my before version https://github.com/benaadams/coreclr/tree/orig-tp) 

Before - 16 threads and 256 connections (16 Piplined)
```
Requests/sec: 146634.07
Requests/sec: 148249.06
Requests/sec: 148670.75
Requests/sec: 147744.71
Requests/sec: 146830.95
```
After
```
Requests/sec: 149823.42
Requests/sec: 148628.71
Requests/sec: 143656.47
Requests/sec: 159852.75
Requests/sec: 150972.37
```
Before - 16 threads and 256 connections
```
Requests/sec:  86273.51
Requests/sec:  84529.18
Requests/sec:  85615.19
Requests/sec:  82796.65
Requests/sec:  87153.29
Requests/sec:  82744.33
```
After
```
Requests/sec:  81118.31
Requests/sec:  85218.90
Requests/sec:  85501.99
Requests/sec:  76305.53
Requests/sec:  75538.75
Requests/sec:  83900.66
```
The CPU is under half utilised, very poor network speed; its a factor of 10 out than what I was expecting; there's no real Kernel time use so no sleeps kicking in.

**Comment by benaadams** [comment]
Eurgh... just realised what's wrong. Didn't change it to Server GC 😱 

Will try again in the morning.

**Comment by benaadams** [comment]
Sadly tore it all down before I realised I'd seen this before and knew exactly what it was...

**Comment by benaadams** [comment]
Something is wacked somewhere... 

20 threads and 1000 connections, pipelined 16

Kestrel 1.1.0; Microsoft.NETCore.App 1.1.0 (Nuget)
```
Requests/sec: 1,935,042.62
```
Kestrel 1.2.0, netcoreapp2.0, coreclr old TP (Coreclr build, Myget corefx dotnetdiv, Myget Kestrel aspnet)
```
Requests/sec: 453,174.77
```
Kestrel 1.2.0, netcoreapp2.0, coreclr new TP (Coreclr build, Myget corefx dotnetdiv, Myget Kestrel aspnet)
```
Requests/sec: 464,816.65
```
Using Kestrel and corefx from mygets (coreclr from build); but somethings off, might have got a checked jit in or a debug build somewhere.

So ignore all the results so far; and I'm going to have to build everything from source all the way to the end to see if there really is a problem or something slow crept in... Might take a while 😭

**Comment by benaadams** [comment]
Isn't my coreclr build, fresh release still has issue and get slow down just with feeds. Building corefx + tests in release mode.

**Comment by benaadams** [comment]
Still regression with coreclr and corefx from source. Building Kestrel then will get the profilers out...

**Comment by benaadams** [comment]
Copy paste issue; from picking up new config changes 😢

**Comment by benaadams** [comment]
Damn its fast, lastest latest; I'm seeing a 25% gain on 1.1

**Comment by benaadams** [comment]
Omg, looks like its the threadpool... Will post the results in the coreclr PR... 😉

**Comment by benaadams** [comment]
Hmm... its inconsistent; though always on par; so I've had up to 2,559,988 rps for 100% cpu; but also 1,9024,439 at 80%

Players looks like sleep and GC



Sleep chain is



Won't really factor for something doing more

**Comment by benaadams** [comment]
For interest the top own times in coreclr are

**Comment by benaadams** [comment]
Probably going to leave this alone for a while; its been a rollacoster...

**Comment by stephentoub** [comment]
Thank you for testing it out.  All in all, looks like a good improvement :)

**Comment by benaadams** [comment]
My copypasterror was switching on logging; which caused it all to go to hell 😆

**Comment by stephentoub** [comment]
> My copypasterror was switching on logging; which caused it all to go to hell 😆

Hehe, yeah, that would do it. :smile:

**Comment by stephentoub** [comment]
As an aside, in your trace, can you see where those EventSource.Write calls are coming from?  I wouldn't expect to see those show up in a trace like this, unless this tool turn on ETW?

**Comment by benaadams** [comment]
Already made a PR for them :) https://github.com/aspnet/Hosting/pull/931

**Comment by stephentoub** [comment]
You're way ahead of me then :)

**Comment by benaadams** [comment]
There's a whole bunch of little things, ArraySegment .ctor doesn't inline which I'm going to look at.

I can share the trace if its helpful?

**Comment by stephentoub** [comment]
Sure, thanks

**Comment by benaadams** [comment]
Has all the dlls and pdbs https://aoa.blob.core.windows.net/aspnet/RawBuild.zip (105MB)

Trace is the `HttpTest.dtp` file for dotTrace https://www.jetbrains.com/profiler/

**Comment by benaadams** [comment]
Actual code, though it isn't terribly exciting https://aoa.blob.core.windows.net/aspnet/HttpTest.zip

Plaintext benchmark without all the extras

**Comment by benaadams** [comment]
SpinWait->SpinOnce->NextSpinWillYield->PlatformHelper.IsSingleProcessor->ProcessorCount == 1

Processor count could maybe be cached in the SpinWait? Though its fiddling at the edges :)

**Comment by stephentoub** [comment]
> Processor count could maybe be cached in the SpinWait?

Potentially.  You could play with it.  Keep in mind, though, that the purpose of SpinWait is to burn time, so decreasing the work it does may actually be harmful to its usage.

**Comment by stephentoub** [comment]
> Trace is the HttpTest.dtp 

Thanks, Ben. I grabbed the trace.

**Pull request merged**